logstash-core/lib/logstash/agent.rb

```diff
@@ -33,7 +33,7 @@ def execute
   end # def execute
 
   def add_pipeline(pipeline_id, config_str, settings = {})
-    @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings)
+    @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings.merge(:pipeline_id => pipeline_id))
```

Comment:
One thing to note is I made the pipeline ID part of the thread names so it looks nice in VisualVM!